x/cash/msg_test.go

```diff
@@ -1,138 +1,149 @@
 package cash
 
 import (
-	"strings"
+	"crypto/rand"
 	"testing"
 
 	"github.com/iov-one/weave"
 	coin "github.com/iov-one/weave/coin"
 	"github.com/iov-one/weave/errors"
-	"github.com/stretchr/testify/assert"
 )
 
 func TestValidateSendMsg(t *testing.T) {
-	empty := new(SendMsg)
-	assert.Equal(t, pathSendMsg, empty.Path())
-	assert.Error(t, empty.Validate())
-
-	addr := weave.NewAddress([]byte{1, 2})
-	addr2 := weave.NewAddress([]byte{3, 4})
-	addr3 := weave.NewAddress([]byte{5, 6})
-
-	pos := coin.NewCoin(10, 0, "FOO")
-	noSrc := &SendMsg{
-		Amount: &pos,
-		Dest:   addr,
-	}
-	err := noSrc.Validate()
-	assert.Error(t, err)
-	assert.True(t, errors.ErrInvalidInput.Is(err))
-
-	// add a default source, so it validates
-	good := noSrc.DefaultSource(addr2)
-	assert.EqualValues(t, addr2, good.GetSrc())
-	assert.NoError(t, good.Validate())
-
-	// don't change source if already set
-	good2 := good.DefaultSource(addr3)
-	assert.EqualValues(t, addr2, good2.GetSrc())
-	assert.NoError(t, good2.Validate())
-
-	// try various error coniditons by modifying a good state
-	good2.Dest = []byte{1, 2, 3}
-	assert.Error(t, good2.Validate())
-
-	// test memo length
-	good3 := noSrc.DefaultSource(addr3)
-	assert.NoError(t, good3.Validate())
-	good3.Memo = "kfjuhewiufhgqwegf"
-	assert.NoError(t, good3.Validate())
-	good3.Memo = strings.Repeat("foo", 300)
-	err = good3.Validate()
-	assert.Error(t, err)
-	assert.True(t, errors.ErrInvalidState.Is(err))
-
-	// test ref length
-	good3.Memo = "short"
-	good3.Ref = []byte{1, 2, 3, 4, 5}
-	assert.NoError(t, good3.Validate())
-	good3.Ref = make([]byte, 68)
-	err = good3.Validate()
-	assert.Error(t, err)
-	assert.True(t, errors.ErrInvalidState.Is(err))
-
-	neg := coin.NewCoin(-3, 0, "FOO")
-	minus := &SendMsg{
-		Amount: &neg,
-		Dest:   addr2,
-		Src:    addr3,
+	addr1 := randomAddr(t)
+	addr2 := randomAddr(t)
+
+	cases := map[string]struct {
+		msg     weave.Msg
+		wantErr *errors.Error
+	}{
+		"success": {
+			msg: &SendMsg{
+				Amount: coin.NewCoinp(10, 0, "FOO"),
+				Dest:   addr1,
+				Src:    addr2,
+				Memo:   "some memo message",
+				Ref:    []byte("some reference"),
+			},
+			wantErr: nil,
+		},
+		"success with minimal amount of data": {
+			msg: &SendMsg{
+				Amount: coin.NewCoinp(10, 0, "FOO"),
+				Dest:   addr1,
+				Src:    addr2,
+			},
+			wantErr: nil,
+		},
+		"empty message": {
+			msg:     &SendMsg{},
+			wantErr: errors.ErrInvalidAmount,
+		},
+		"missing source": {
+			msg: &SendMsg{
+				Amount: coin.NewCoinp(10, 0, "FOO"),
+				Dest:   addr1,
+			},
+			wantErr: errors.ErrEmpty,
+		},
+		"missing destination": {
+			msg: &SendMsg{
+				Amount: coin.NewCoinp(10, 0, "FOO"),
+				Src:    addr2,
+			},
+			wantErr: errors.ErrEmpty,
+		},
+		"reference too long": {
+			msg: &SendMsg{
+				Amount: coin.NewCoinp(10, 0, "FOO"),
+				Dest:   addr1,
+				Src:    addr2,
+				Ref:    []byte("aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaB"),
```

Comment:
q: how many chars are this? I find this harder to read than calling some function that creates the string. for example bigString(128)
q: is there any reason for `B` at the end?

Reply:
I have updated to use `strings.Repeat` to make the characters count easier.